Fix ZooKeeper backend suffix backslash key

chenchun opened the pull request:
Signed-off-by: Chun Chen <ramichen@tencent.com>

Keys with suffix backslash is not allowed for zookeeper.
```
chenchun@chenchun-T450:/usr/local/zookeeper-3.4.6$ bin/zkCli.sh
...
[zk: localhost:2181(CONNECTED) 0] ls /foo
[]
[zk: localhost:2181(CONNECTED) 1] ls /foo/
Command failed: java.lang.IllegalArgumentException: Path must not end with / character
[zk: localhost:2181(CONNECTED) 2] create /foo1/ ""
Command failed: java.lang.IllegalArgumentException: Path must not end with / character
[zk: localhost:2181(CONNECTED) 3] create /foo1 "" 
Created /foo1
[zk: localhost:2181(CONNECTED) 4] create //foo2 "" 
Command failed: java.lang.IllegalArgumentException: Invalid path string "//foo2" caused by empty node name specified @1
```

chenchun commented:
Rethinked of this, it seems this is only a problem of zookeeper. Updated the patch to only refactor keys of zookeeper. I made this PR cause https://github.com/docker/libnetwork/blob/master/datastore/datastore.go#L73 is using keys with suffix backslash.

chenchun commented:
Rebased to fix `List` function.

abronan commented:
Hi @chenchun, thanks again for the contribution and sorry for the delay. Still unsure about that, an easier fix is to remove the line in `libnetwork` that appends the `/` character at the end. Because this is the *supported* format:

`dir1/dir2/key`

I agree that the key parsing is lacking though and that this needs to be better handled and *flexible*. I'd probably just wait a little bit for the BoltDB support and see what's the best plan to refactor the whole key parsing. Maybe including the key parsing portion in `datastore.go` from **libnetwork** could be a good idea.

abronan commented:
Hmm sorry didn't see the related PR on libnetwork's side, will catch up and review.

chenchun commented:
> Maybe including the key parsing portion in datastore.go from libnetwork could be a good idea.

I didn't change it in libnetwork because I think it is good to handle the key forms of different  backends and offer users a unified format in libkv since keys with tailing `/` are supported both by etcd and consul. It's like the way that libkv handles key not found.

> Because this is the supported format:
> `dir1/dir2/key`

I can't find any docs talking about this. In fact the same situation exists for keys with `/` prefix. etcd and zookeeper supports keys with `/` prefix, but not consul.

abronan commented on testutils/utils.go:
```diff
@@ -24,44 +24,46 @@ func RunTestStore(t *testing.T, kv store.Store, backup store.Store) {
-	assert.False(t, exists)
+	for key, value := range map[string][]byte{
+		"testfoo":  []byte("bar"),
+		"testfoo/": []byte("bar"),
```
We can probably add two more examples of keys here :)
```
testfoo/testbar/
testfoo/testbar/testfoobar
```

chenchun replied:
Added in the latest patch.

abronan commented:
@chenchun Sorry for the delay, LGTM as it is, can you add the two more examples to the tests? Thanks! 

I'm not fond of the way we parse keys today but this is good until we figure out a better way, extensible to other stores. Bonus: this works with boltDB so we good :)

abronan commented:
@chenchun Thanks! I'm sorry about this but can you put the zookeeper patch (fixing `[]byte{1}`) alone in a separate PR? I need to test that independently from the rest to see if this was effectively the root cause of the issue (or if we go back to the state where some entries are missed in swarm discovery).

The `Cleanup` function in tests is much needed though! :smile:

chenchun commented:
@abronan I found lots of keys in the tests are not deleted and can mix up next build with Travis CI. 
So I refactored keys in tests and cleaned them after tests.

abronan commented:
@chenchun :+1:

chenchun commented:
> I'm sorry about this but can you put the zookeeper patch (fixing []byte{1}) alone in a separate PR?

The tests of zookeeper will fail if I do that.

abronan commented:
Oh ok then :) I'll take some time to test it thoroughly. Trying to understand why putting a placeholder value on a directory could randomly impact `GetChildrens` and make an entry vanish like this. Need to test it with swarm and delete/add nodes to see if this happens again. I remember this happening very rarely and I see no corelation between the value we put in the directories created in `fullpath` and the actual value at `key` (ie. putting a value here is useless but adding one shouldn't impact `GetChildrens` either).

abronan commented on store/etcd/etcd.go:
```diff
@@ -160,6 +160,9 @@ func (s *Etcd) Put(key string, value []byte, opts *store.WriteOptions) error {
 func (s *Etcd) Delete(key string) error {
 	_, err := s.client.Delete(store.Normalize(key), false)
+	if isKeyNotFoundError(err) {
+		return nil
```
Shouldn't this return `store.ErrKeyNotFound` instead of `nil`?

chenchun replied:
Should I change `boltdb`/`consul` as well?

abronan replied:
I'd say at least for Consul and leave the tests as it is so that we don't enforce the same error on `boltdb`. We can do `boltdb` in another PR to fix it and add the tests on `store.ErrKeyNotFound` :)

chenchun replied:
Fixed.

abronan commented on store/etcd/etcd.go:
```diff
@@ -366,6 +369,9 @@ func (s *Etcd) List(directory string) ([]*store.KVPair, error) {
 func (s *Etcd) DeleteTree(directory string) error {
 	_, err := s.client.Delete(store.Normalize(directory), true)
+	if isKeyNotFoundError(err) {
+		return nil
```
same here

abronan commented on store/zookeeper/zookeeper.go:
```diff
@@ -127,13 +118,16 @@ func (s *Zookeeper) Put(key string, value []byte, opts *store.WriteOptions) erro
-	err := s.client.Delete(store.Normalize(key), -1)
+	err := s.client.Delete(s.normalize(key), -1)
+	if err == zk.ErrNoNode {
+		return nil
```
same here

abronan commented on store/zookeeper/zookeeper.go:
```diff
@@ -263,6 +257,9 @@ func (s *Zookeeper) List(directory string) ([]*store.KVPair, error) {
 	pairs, err := s.List(directory)
 	if err != nil {
+		if err == store.ErrKeyNotFound {
+			return nil
```
here as well :)

abronan commented:
@chenchun I tested the PR by updating Godeps in swarm against zookeeper `3.4.5`/`3.4.6` and can't reproduce the bug on the `WatchTree` weird behavior. Nice catch! Hope this was not me being insanely lucky but I tested this on thousands of watch events :smile: 

I left a few comments on the key exist checks. Would be nice to just add a few lines of tests in `DeleteTree` and `Delete`, checking a delete on a key that does not exist (by checking `store.ErrKeyNotFound`)

chenchun commented:
Thanks for testing it. Updated the PR.

abronan commented:
LGTM

Thanks @chenchun!
